### ISSUE TITLE
RI layer and OED modules

### ISSUE THREAD

**Issue by sr-murthy** [OasisLMF/OasisLMF#185]
@mpinkerton-oasis There seems to be a number of problems with these RI layer and OED modules - after the latest commits there are still issues in running some of the test cases  in the RTT test notebook, e.g. this is the latest one when trying to run FM3

    ---------------------------------------------------------------------------
    TypeError                                 Traceback (most recent call last)
    <ipython-input-7-7398a681cd68> in <module>
         21 (ri_layers, xref_descriptions) = generate_oasis_files(
         22     direct_oasis_files_dir, output_dir,
    ---> 23     srcexptocan_trans_fp, srcacctocan_trans_fp)
         24 loss_factor = 0.5
         25 net_losses = generate_losses(

    ~/src/oasislmf/oasislmf/utils/deterministic_loss.py in generate_oasis_files(input_dir, output_dir, srcexptocan_trans_fp, srcacctocan_trans_fp)
         72     if do_reinsurance:
         73         oed_validator = oed.OedValidator()
    ---> 74         (is_valid, error_msgs) = oed_validator.validate(ri_info_df, ri_scope_df)
         75         if not is_valid:
         76             print("Validation Failed:")

    ~/src/oasislmf/oasislmf/model_preparation/oed.py in validate(self, ri_info_df, ri_scope_df)
        161                 error_msg))
        162 
    --> 163         for inuring_priority in range(1, ri_info_df['InuringPriority'].max() + 1):
        164             inuring_priority_ri_info_df = ri_info_df[ri_info_df.InuringPriority == inuring_priority]
        165             if inuring_priority_ri_info_df.empty:

    TypeError: 'float' object cannot be interpreted as an integer

The problem here is possibly that that the `InuringPriority` column at this point (line 163) contains floats, so that `range` will not work - it only accepts integers. This makes the `validate` method a bit useless.

The Pandas code in the RI layer and OED modules is also faulty in the way that frame and series attributes are accessed, e.g. `location.LocGroup` will only work if the series object has the `LocGroup` column, which will only be the case if the location file from which frame was created has the `LocGroup` column. Unless you've set column defaults in some other way, the correct approach 
is to use

    location.get('LocGroup', <default value>)

I've made a fix for this, by replacing all occurrences of direct attribute access with the use of `get`, but you still have some problems in `validate` and elsewhere.

**Comment by mpinkerton-oasis** [comment]
The inuring priority should be an integer value. I think the validation rule should be changed be int_only, not int_or_float.

As an aside, Simplitium are working on a JSON Schema for OED which will encapsulate all of the validation logic. I think we should think about using this rather than implementing complex validation logic in the code.

**Comment by sr-murthy** [comment]
OK, but I think a fix is neeed here in your `oed.py` module

    ~/src/oasislmf/oasislmf/model_preparation/oed.py in validate(self, ri_info_df, ri_scope_df)
    161                 error_msg))
    162 
--> 163         for inuring_priority in range(1, ri_info_df['InuringPriority'].max() + 1):
    164             inuring_priority_ri_info_df = ri_info_df[ri_info_df.InuringPriority == inuring_priority]
    165             if inuring_priority_ri_info_df.empty:

when you receive that `ri_info_df` frame, you need to either explicitly set the column types appropriately, or catch exceptions in case of unexpected types. Neither is happening here.

As for the schema, fine, but only this RI layer implementation is using OED validation - there is no OED validation going on elsewhere.

**Comment by mpinkerton-oasis** [comment]
It also sounds like there is a possible issue with your test data if the inuring_priority is a float.

**Comment by sr-murthy** [comment]
No. I think you're confusing the FM acceptance test failures that were caused by your changes in the deterministic loss utils module, which I've now fixed, with the error I posted above - that was from the RTT repository. The test data isn't mine. The problem in any case is not the data, it is the code in the OED module.

**Comment by sr-murthy** [comment]
FYI The FM acceptance test cases have no RI component - there is no RI. These are just direct losses.

**Comment by mpinkerton-oasis** [comment]
In that case the error is that do_reinsurance should be false and the reinsurance code never hit. What exactly are you running to generate the error?

**Comment by sr-murthy** [comment]
No that is incorrect - the errors appears to be in the OED `load_oed_dfs` method that you've added as part of your changes to the `generate_oasis_files` method in the deterministic loss utils module. You have this statement in `generate_oasis_files`

    ...
     (ri_info_df, ri_scope_df, do_reinsurance) = oed.load_oed_dfs(input_dir)
     ...

Now, `input_dir` will take the value of whatever input directory was passed in the call to `generate_oasis_files`. As this method is used to generate the direct Oasis files in the RTT test notebook (this step used to be helpfully numbered #7, but I see this has been removed) from a given test case or exampe directory, e.g. `ftest/fm3`, the value of `input_dir` will be say `ftest/fm3` when running FM3. One would think that this method `oed.oed_load_dfs` would be able to determine whether the input directory contained any RI assets, and if not would return `False` for `do_reinsurance` for this example.

The error that I posted above is happening in the `load_oed_dfs` method because it probably mistakenly returning `True` when the `input_dir` is FM3, which means the value of `do_reinsurance` in `generate_oasis_files` is `True`, and so it makes the call to `validate`, which is where the error occurs when trying to process the RI info DF, which doesn't exist for FM3.

So there appears to be a logic error in OED, and I would say that this is something that should be picked during static testing, but it doesn't appear to have any static tests, i.e. I don't see any `test_oed.py`.

I would say that the RI layer and OED modules are the least robust parts of the MDK, and most error-prone.

**Comment by sr-murthy** [comment]
No that is incorrect - the errors appears to be in the OED `load_oed_dfs` method that you've added as part of your changes to the `generate_oasis_files` method in the deterministic loss utils module. You have this statement in `generate_oasis_files`

    ...
     (ri_info_df, ri_scope_df, do_reinsurance) = oed.load_oed_dfs(input_dir)
     ...

Now, `input_dir` will take the value of whatever input directory was passed in the call to `generate_oasis_files`. As this method is used to generate the direct Oasis files in the RTT test notebook (this step used to be helpfully numbered #7, but I see this has been removed) from a given test case or exampe directory, e.g. `ftest/fm3`, the value of `input_dir` will be say `ftest/fm3` when running FM3. One would think that this method `oed.oed_load_dfs` would be able to determine whether the input directory contained any RI assets, and if not would return `False` for `do_reinsurance` for this example.

The error that I posted above is happening in the `load_oed_dfs` method because it is probably mistakenly returning `True` when the `input_dir` is FM3, which means the value of `do_reinsurance` in `generate_oasis_files` is `True`, and so it makes the call to `validate`, which is where the error occurs when trying to process the RI info DF, which doesn't exist for FM3.

So there appears to be a logic error in OED, and I would say that this is something that should be picked up during static testing, but it doesn't appear to have any static tests, i.e. I don't see any `test_oed.py`.

I would say that the RI layer and OED modules are the least robust parts of the MDK, and most error-prone.
